nifi-extension-bundles/nifi-flow-registry-client-bundle/nifi-flow-registry-client-services/src/main/java/org/apache/nifi/registry/flow/NifiRegistryFlowRegistryClient.java

```diff
@@ -149,7 +149,7 @@ public boolean isStorageLocationApplicable(final FlowRegistryClientConfiguration
     }
 
     @Override
-    public Set<FlowRegistryBucket> getBuckets(final FlowRegistryClientConfigurationContext context) throws FlowRegistryException, IOException {
+    public Set<FlowRegistryBucket> getBuckets(final FlowRegistryClientConfigurationContext context, final String branch) throws FlowRegistryException, IOException {
```

Comment:
I would say it is really up to the implementation to decide. We could have all the methods in NiFi registry check the incoming branch argument, and validate that it equals the same name as the default branch, but I don't think it is providing much value since none of the methods are using the branch.

Can you clarify what you mean about `getFlowContents` ? I believe that method is unchanged in the NiFi Registry implementation.

Reply:
I wanted to point [here](https://github.com/bbende/nifi/blob/NIFI-13120/nifi-extension-bundles/nifi-flow-registry-client-bundle/nifi-flow-registry-client-services/src/main/java/org/apache/nifi/registry/flow/NifiRegistryFlowRegistryClient.java#L242-L247) as an example. The point was, here (even though it will aways be the default one) the branch is handled as it coukld be different and we might expect something else from the argument then the default.

Thus it looks like it could have different value, which (in case of this particular implementation) is not true at this point. I think this "proper handling" can be misleading and might result unrealistic responses: it can return a snapshot with proper content but having metadata pointing to an arbitrary (non-existing) branch.

Reply:
Are you saying that you think we should check the `branch` argument in all of the methods in `NiFiRegistryFlowRegistryClient` to validate that it matches the default branch?

The only way this could happen is someone making API calls outside of the UI where they could submit a request for save/import with a branch name that wasn't the default branch NiFi Registry is expecting, but also it would just be ignored by all of the methods, so it wouldn't cause any issues. 

One other note... I did not add `branch` to the NiFi Registry specific data model class like `VersionedFlow`, `VersionedFlowSnapshot`, and `VersionedFlowSnapshotMetadat`, because its not used for anything on registry side. So in your example, if someone calls `getFlowContents` with `branch` as `dev`, they will get back a null branch in the response since there wasn't a branch in the content returned from registry.

Happy to improve something here, just want to make sure I'm following what you are suggesting.

Reply:
I have no strong feelings about it but I see some possible benefit:
- 1. It looks generally a good approach to not fully rely on client side data validation (due to possible changes there, etc.)
- 2I did see some usage where NiFi API was used directly from scripts/automazation. While I do not think that this is the intended way (however by supporting "Request Auth Token" this seems inevidable) there is a possibility for this and I would be on the safe side even if the current version of the code could not done real harm with using non-existing branch.

I am not sure which is the more forwarding way to tackle this. We can either add validation (it looks a bit much future proof but might be misleading as it would look like this is an argument we really consider here) or just enforce the default value (which comes with some hardwired thing, generally to avoid) but I think it would be safer to do some steps towards some of these directions.

Reply:
Thanks for clarifying. I went ahead and added validation within the NiFi Registry client that enforces that if any method receives a non-null branch, then it must match the default branch.